### DISCUSSION TITLE
Installing or upgrading p11-kit hangs on 'make check'

### DISCUSSION THREAD

**Discussion by plexilcrf** [Homebrew/discussions#2814] [Getting started] [answered]
In macOS Big Sur 11.6.2, on an Intel MacBook Pro (16-inch, 2019):

I recently had to upgrade my macOS installation (long story, it's a US Gov't site, we're always a year behind due to security policies) and of course my Homebrew installation, once I had a working set of build tools again. Every time I tried 'brew upgrade outdated', it would hang upgrading p11-kit, in 'make check'. As far as I could tell, the upgrade process was doing nothing at this stage - not spinning, just waiting.

So I cleaned up a ton of redundant packages, both in Homebrew and M*cP*rts. Everything was fine ... until I realized I'd deleted my Emacs.app installation.

Go to install the Homebrew 'emacs' package, and what should turn up as a dependency? You guessed it, p11-kit. And so the installation is hanging at 'make check'.

What could cause this phase of the installation to hang? Any clues would be welcomed. Thank you.

**Comment by carlocab** [comment, marked as answer]
It's hard to say without more information. I take you're unable to install Homebrew into `/usr/local`?

You could probably get away with commenting out or deleting the `make check` line in `p11-kit.rb` You can do this with `brew edit p11-kit`, but be sure to `git -C "$(brew --repo homebrew/core)" restore .` after.

It'll probably work fine; we've been building `p11-kit` without running the test suite for a while with no complaints. Except given what it's used for, it seems more prudent to run it when building our binary packages.

**Reply by plexilcrf** [reply]
Correct, I'm unable to install into /usr/local without going to a sysadmin, I have Homebrew installed in my home directory so I don't have to ask "Mother may I?" all the time.

Thank you for the tip, I seem to have gotten over that hurdle!

**Reply by gromgit** [reply]
Sidenote: if by "Mother may I?" you mean "Mother may I _have root privileges_?", only the Homebrew installer needs that, and only once during the installation process. Thereafter, you can install Homebrew formulae via _precompiled and tested bottles_ to your heart's content as yourself, no root or building from source needed.

**Reply by Yann-R** [reply]
@carlocab 
> You could probably get away with commenting out or deleting the make check line in p11-kit.rb You can do this with brew edit p11-kit, but be sure to git -C "$(brew --repo homebrew/core) restore ." after.

Thanks for the tips. 👍
Note that, to avoid getting the error "No such file of directory", the last command to restore the edited file must be invoked from the root of the local repo directory (given by `brew --repo`)

**Reply by carlocab** [reply]
> Note that, to avoid getting the error "No such file of directory", the last command to restore the edited file must be invoked from the root of the local repo directory (given by `brew --repo`)

That's not necessary, since the `-C` flag is already supposed to adjust this. There was a typo in the command I suggested, though. It's fixed now.

**Reply by Yann-R** [reply]
Great!
Many thanks.

**Comment by Yann-R** [comment]
Hi all,

(the above tip to comment out the final tests that break the build is working for me)

But, I encounter the same problem to build from source with the full (and unmodified) formula:
- when building `p11-kit`from source on my Intel-based system (under Big Sur) with non standard homebrew prefix `/opt/brew`
- the command `brew upgrade p11-kit` remains stuck forever on the last `make check`: 
```
==> Upgrading 1 outdated package:
p11-kit 0.24.0_2 -> 0.24.1
[...]
==> ./configure --prefix=/opt/brew/Cellar/p11-kit/0.24.1 --sysconfdir=/opt/brew/etc --with-module-config=/opt/brew/etc/pkcs11/modules --with-trust-paths=/opt/brew/etc/ca-certificates/cert.pem
==> make
==> make check
```


> @carlocab: It's hard to say without more information.

To give more information, it appears that `make check` passes with success only 755 tests of 757 and then the process remains stuck, without stopping:
- I would expect `brew` to stop and display the error
- And of course, I'd like to find the correction to make these 2 failed tests pass correctly

Here follows the detailed info from brew with debug mode and interactive shell:
```
$ brew upgrade -d p11-kit
[...]
==> ./configure --prefix=/opt/brew/Cellar/p11-kit/0.24.1 --sysconfdir=/opt/brew/etc --with-module-config=/opt/brew/etc/pkcs11/modules --with-trust-paths=/opt/brew/etc/ca-certificates/cert.pem
==> make
==> make check
^C <-- stopped by ctrl-C
Interrupt: 
1. raise
2. backtrace
3. shell
Choose an action: 3
When you exit this shell, you will return to the menu.
[computer:tmp/p11-kit-20220421-75765-1gyzev9/p11-kit-0.24.1] me$ make check
[...]
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: test-hash 1 /hash/murmur3
[...]
SKIP: test-compat 2 /compat/getprogname # SKIP cannot perform getprogname test
PASS: test-constants 1 /constants/types
PASS: test-constants 2 /constants/classes
[...]
ERROR: p11-kit/test-server.sh - too few tests run (expected 4, got 1)
PASS: test-token 1 /token/load
PASS: test-token 2 /token/flags
PASS: test-token 3 /token/path
PASS: test-token 4 /token/label
PASS: test-token 5 /token/slot
[...]
============================================================================
Testsuite summary for p11-kit 0.24.1
============================================================================
# TOTAL: 757
# PASS:  755
# SKIP:  1
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 1
============================================================================
See ./test-suite.log
Please report to https://github.com/p11-glue/p11-kit/issues
============================================================================
make[4]: *** [test-suite.log] Error 1
make[3]: *** [check-TESTS] Error 2
make[2]: *** [check-am] Error 2
make[1]: *** [check-recursive] Error 1
make: *** [check] Error 2
```
And at this point, the process is stuck forever as described above.
Note the two problems during the tests (one SKIP and one ERROR)
Don't know why this occur here and not when the bottle for brew is built.

**Reply by carlocab** [reply]
Why are you building from source?

**Reply by Yann-R** [reply]
My brew builds (often) from source since I am not using the default prefix (but `/opt/brew` as mentioned at the beginning of my report) because I install also other sources of software than brew formulae or casks: from my early years with `brew` I preferred not to mix brewed (and well-ordered stuff) with other pieces of software dropped without much precaution by third-party packages and not always resulting in good cohabitation (sometimes overwriting brewed stuff), not very encouraging when you want your system as clean as possible.

I'm practicing it like some other people (e.g. [here](https://github.com/Homebrew/brew/issues/9130)), while being aware that using another location than the default one for `brew` is possible but (as documented [here](https://docs.brew.sh/Installation#untar-anywhere)) not encouraged: "Just extract (or git clone) Homebrew wherever you want. [...] However do yourself a favour and use the installer to install to the default prefix. Some things may not build when installed elsewhere."

All over these years, and several Macs, I had only less than 5 formulae not working among around 500 installed ones. Each time I encountered one, I tried to find a solution myself or to cooperate like here to correct the formula.

In this context, `brew` builds from source when it had to do so (I never force it). For instance, concerning `p11-kit`:
```
Warning: Building p11-kit from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /opt/brew/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /opt/brew)
```

Then, for the concerned problem, I really don't understand why 2 tests among 757 are failing when compiling and checking the embedded p11-kit tests from the source and not when the bottle was built.

**Reply by jslay-excella** [reply]
Hitting the slightly different issue (doesn't hang, but fails `make check`), but different output.

```
============================================================================
Testsuite summary for p11-kit 0.24.1
============================================================================
# TOTAL: 759
# PASS:  757
# SKIP:  1
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to https://github.com/p11-glue/p11-kit/issues
============================================================================
```

Also, not uncommon to have to build from source because of the prefix being set for a lot of workplace laptops. 

```
Warning: Building p11-kit from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /Users/$USER/homebrew/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /Users/$USER/homebrew)
```

I disabled tests as suggested with `brew edit p11-kit` and moved on. 

Failed test output (probably not relevant for Homebrew):

```
FAIL: p11-kit/test-messages.sh 1 /messages/return-code
# p11-kit/test-messages.sh: 1,81d0
# p11-kit/test-messages.sh: < CKR_CANCEL: The operation was cancelled
# p11-kit/test-messages.sh: < CKR_FUNCTION_CANCELED: The operation was cancelled
# p11-kit/test-messages.sh: < CKR_HOST_MEMORY: Insufficient memory available
# p11-kit/test-messages.sh: < CKR_SLOT_ID_INVALID: The specified slot ID is not valid
# p11-kit/test-messages.sh: < CKR_GENERAL_ERROR: Internal error
# p11-kit/test-messages.sh: < CKR_FUNCTION_FAILED: The operation failed
# p11-kit/test-messages.sh: < CKR_ARGUMENTS_BAD: Invalid arguments
# p11-kit/test-messages.sh: < CKR_NEED_TO_CREATE_THREADS: The module cannot create needed threads
# p11-kit/test-messages.sh: < CKR_CANT_LOCK: The module cannot lock data properly
# p11-kit/test-messages.sh: < CKR_ATTRIBUTE_READ_ONLY: The field is read-only
# p11-kit/test-messages.sh: < CKR_ATTRIBUTE_SENSITIVE: The field is sensitive and cannot be revealed
# p11-kit/test-messages.sh: < CKR_ATTRIBUTE_TYPE_INVALID: The field is invalid or does not exist
# p11-kit/test-messages.sh: < CKR_ATTRIBUTE_VALUE_INVALID: Invalid value for field
# p11-kit/test-messages.sh: < CKR_DATA_INVALID: The data is not valid or unrecognized
# p11-kit/test-messages.sh: < CKR_DATA_LEN_RANGE: The data is too long
# p11-kit/test-messages.sh: < CKR_DEVICE_ERROR: An error occurred on the device
# p11-kit/test-messages.sh: < CKR_DEVICE_MEMORY: Insufficient memory available on the device
# p11-kit/test-messages.sh: < CKR_DEVICE_REMOVED: The device was removed or unplugged
# p11-kit/test-messages.sh: < CKR_ENCRYPTED_DATA_INVALID: The encrypted data is not valid or unrecognized
# p11-kit/test-messages.sh: < CKR_ENCRYPTED_DATA_LEN_RANGE: The encrypted data is too long
# p11-kit/test-messages.sh: < CKR_FUNCTION_NOT_SUPPORTED: This operation is not supported
# p11-kit/test-messages.sh: < CKR_KEY_HANDLE_INVALID: The key is missing or invalid
# p11-kit/test-messages.sh: < CKR_KEY_SIZE_RANGE: The key is the wrong size
# p11-kit/test-messages.sh: < CKR_KEY_TYPE_INCONSISTENT: The key is of the wrong type
# p11-kit/test-messages.sh: < CKR_KEY_NOT_NEEDED: No key is needed
# p11-kit/test-messages.sh: < CKR_KEY_CHANGED: The key is different than before
# p11-kit/test-messages.sh: < CKR_KEY_NEEDED: A key is needed
# p11-kit/test-messages.sh: < CKR_KEY_INDIGESTIBLE: Cannot include the key in the digest
# p11-kit/test-messages.sh: < CKR_KEY_FUNCTION_NOT_PERMITTED: This operation cannot be done with this key
# p11-kit/test-messages.sh: < CKR_KEY_NOT_WRAPPABLE: The key cannot be wrapped
# p11-kit/test-messages.sh: < CKR_KEY_UNEXTRACTABLE: Cannot export this key
# p11-kit/test-messages.sh: < CKR_MECHANISM_INVALID: The crypto mechanism is invalid or unrecognized
# p11-kit/test-messages.sh: < CKR_MECHANISM_PARAM_INVALID: The crypto mechanism has an invalid argument
# p11-kit/test-messages.sh: < CKR_OBJECT_HANDLE_INVALID: The object is missing or invalid
# p11-kit/test-messages.sh: < CKR_OPERATION_ACTIVE: Another operation is already taking place
# p11-kit/test-messages.sh: < CKR_OPERATION_NOT_INITIALIZED: No operation is taking place
# p11-kit/test-messages.sh: < CKR_PIN_INCORRECT: The password or PIN is incorrect
# p11-kit/test-messages.sh: < CKR_PIN_INVALID: The password or PIN is invalid
# p11-kit/test-messages.sh: < CKR_PIN_LEN_RANGE: The password or PIN is of an invalid length
# p11-kit/test-messages.sh: < CKR_PIN_EXPIRED: The password or PIN has expired
# p11-kit/test-messages.sh: < CKR_PIN_LOCKED: The password or PIN is locked
# p11-kit/test-messages.sh: < CKR_SESSION_CLOSED: The session is closed
# p11-kit/test-messages.sh: < CKR_SESSION_COUNT: Too many sessions are active
# p11-kit/test-messages.sh: < CKR_SESSION_HANDLE_INVALID: The session is invalid
# p11-kit/test-messages.sh: < CKR_SESSION_READ_ONLY: The session is read-only
# p11-kit/test-messages.sh: < CKR_SESSION_EXISTS: An open session exists
# p11-kit/test-messages.sh: < CKR_SESSION_READ_ONLY_EXISTS: A read-only session exists
# p11-kit/test-messages.sh: < CKR_SESSION_READ_WRITE_SO_EXISTS: An administrator session exists
# p11-kit/test-messages.sh: < CKR_SIGNATURE_INVALID: The signature is bad or corrupted
# p11-kit/test-messages.sh: < CKR_SIGNATURE_LEN_RANGE: The signature is unrecognized or corrupted
# p11-kit/test-messages.sh: < CKR_TEMPLATE_INCOMPLETE: Certain required fields are missing
# p11-kit/test-messages.sh: < CKR_TEMPLATE_INCONSISTENT: Certain fields have invalid values
# p11-kit/test-messages.sh: < CKR_TOKEN_NOT_PRESENT: The device is not present or unplugged
# p11-kit/test-messages.sh: < CKR_TOKEN_NOT_RECOGNIZED: The device is invalid or unrecognizable
# p11-kit/test-messages.sh: < CKR_TOKEN_WRITE_PROTECTED: The device is write protected
# p11-kit/test-messages.sh: < CKR_UNWRAPPING_KEY_HANDLE_INVALID: Cannot import because the key is invalid
# p11-kit/test-messages.sh: < CKR_UNWRAPPING_KEY_SIZE_RANGE: Cannot import because the key is of the wrong size
# p11-kit/test-messages.sh: < CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT: Cannot import because the key is of the wrong type
# p11-kit/test-messages.sh: < CKR_USER_ALREADY_LOGGED_IN: You are already logged in
# p11-kit/test-messages.sh: < CKR_USER_NOT_LOGGED_IN: No user has logged in
# p11-kit/test-messages.sh: < CKR_USER_PIN_NOT_INITIALIZED: The user's password or PIN is not set
# p11-kit/test-messages.sh: < CKR_USER_TYPE_INVALID: The user is of an invalid type
# p11-kit/test-messages.sh: < CKR_USER_ANOTHER_ALREADY_LOGGED_IN: Another user is already logged in
# p11-kit/test-messages.sh: < CKR_USER_TOO_MANY_TYPES: Too many users of different types are logged in
# p11-kit/test-messages.sh: < CKR_WRAPPED_KEY_INVALID: Cannot import an invalid key
# p11-kit/test-messages.sh: < CKR_WRAPPED_KEY_LEN_RANGE: Cannot import a key of the wrong size
# p11-kit/test-messages.sh: < CKR_WRAPPING_KEY_HANDLE_INVALID: Cannot export because the key is invalid
# p11-kit/test-messages.sh: < CKR_WRAPPING_KEY_SIZE_RANGE: Cannot export because the key is of the wrong size
# p11-kit/test-messages.sh: < CKR_WRAPPING_KEY_TYPE_INCONSISTENT: Cannot export because the key is of the wrong type
# p11-kit/test-messages.sh: < CKR_RANDOM_SEED_NOT_SUPPORTED: Unable to initialize the random number generator
# p11-kit/test-messages.sh: < CKR_RANDOM_NO_RNG: No random number generator available
# p11-kit/test-messages.sh: < CKR_DOMAIN_PARAMS_INVALID: The crypto mechanism has an invalid parameter
# p11-kit/test-messages.sh: < CKR_BUFFER_TOO_SMALL: Not enough space to store the result
# p11-kit/test-messages.sh: < CKR_SAVED_STATE_INVALID: The saved state is invalid
# p11-kit/test-messages.sh: < CKR_INFORMATION_SENSITIVE: The information is sensitive and cannot be revealed
# p11-kit/test-messages.sh: < CKR_STATE_UNSAVEABLE: The state cannot be saved
# p11-kit/test-messages.sh: < CKR_CRYPTOKI_NOT_INITIALIZED: The module has not been initialized
# p11-kit/test-messages.sh: < CKR_CRYPTOKI_ALREADY_INITIALIZED: The module has already been initialized
# p11-kit/test-messages.sh: < CKR_MUTEX_BAD: Cannot lock data
# p11-kit/test-messages.sh: < CKR_MUTEX_NOT_LOCKED: The data cannot be locked
# p11-kit/test-messages.sh: < CKR_FUNCTION_REJECTED: The request was rejected by the user
```

**Comment by TheBuzzer67** [comment]
Same problem on Montery 12.4! Worked with "brew edit p11-kit" but have to do while it is stop! Thanks!

Version du système :	macOS 12.4 (21F79)
Version du noyau :	Darwin 21.5.0

Nom du modèle :	Mac mini
  Identifiant du modèle :	Macmini8,1
  Nom du processeur :	Intel Core i7 6 cœurs
  Vitesse du processeur :	3,2 GHz
  Nombre de processeurs :	1
  Nombre total de cœurs :	6


bash-3.2$ brew install --build-from-source p11-kit
Warning: your HOMEBREW_PREFIX is set to /usr/local but HOMEBREW_CELLAR is set
to /usr/local/Homebrew/Cellar. Your current HOMEBREW_CELLAR location will stop
you being able to use all the binary packages (bottles) Homebrew provides. We
recommend you move your HOMEBREW_CELLAR to /usr/local/Cellar which will get you
access to all bottles."
==> Downloading https://github.com/p11-glue/p11-kit/releases/download/0.24.1/p11-kit-0.24.1.tar.xz
Already downloaded: /Volumes/Macintosh HD2/Users/buzzer/Library/Caches/Homebrew/downloads/eb8fa782e9c255ccaf544d43c0dd61da1343fc7f83b749024108d1671cb78dee--p11-kit-0.24.1.tar.xz
==> ./configure --prefix=/usr/local/Homebrew/Cellar/p11-kit/0.24.1 --sysconfdir=/usr/local/etc --with-module-config=/usr/local/etc/pkcs11/modules --with-trust-paths=/usr/local/etc/ca-certificates/cert.pem
==> make
==> make check
